pkg/chrootarchive/cmd_linux.go

```diff
@@ -0,0 +1,41 @@
+package chrootarchive
+
+import (
+	"os/exec"
+	"syscall"
+
+	"github.com/docker/docker/pkg/mount"
+	"github.com/pkg/errors"
+	"golang.org/x/sys/unix"
+)
+
+func configureSysProc(cmd *exec.Cmd) {
+	cmd.SysProcAttr = &unix.SysProcAttr{
+		Cloneflags: unix.CLONE_NEWNET | unix.CLONE_NEWPID | unix.CLONE_NEWIPC | unix.CLONE_NEWUTS | unix.CLONE_NEWNS,
+		Pdeathsig:  syscall.SIGKILL,
+	}
+}
+
+// As part of setting up this process we created a new mount ns and a new pid ns.
+// This configures the mount ns to take advantage of that to further isolate the process.
+func setupMountNS() error {
+	// Make everything in new ns slave.
+	// Don't use `private` here as this could race where the mountns gets a
+	//   reference to a mount and an unmount from the host does not propagate,
+	//   which could potentially cause transient errors for other operations,
+	//   even though this should be relatively small window here `slave` should
+	//   not cause any problems.
+	if err := mount.MakeRSlave("/"); err != nil {
+		return errors.Wrap(err, "error remounting rootfs as with slave mounts")
+	}
+
+	if err := mount.Unmount("/proc"); err != nil {
+		return errors.Wrap(err, "error unmounting /proc")
+	}
+
+	// Remount /proc so the
+	if err := unix.Mount("proc", "/proc", "proc", 0, "hidepid=2"); err != nil {
```

Comment:
Hmm, and what would be the mount propagation of this new mount? Is it the one of the parent mount? (frankly I can't remember atm)

Maybe it makes sense to add unix.MS_SLAVE flag (since `unix.Mount` does proper thing in this case)?

Reply:
From http://man7.org/linux/man-pages/man7/mount_namespaces.7.html:

       The propagation type assigned to a new mount point depends on the
       propagation type of the parent mount.  If the mount point has a
       parent (i.e., it is a non-root mount point) and the propagation type
       of the parent is MS_SHARED, then the propagation type of the new
       mount is also MS_SHARED.  Otherwise, the propagation type of the new
       mount is MS_PRIVATE.

So the propagation type will be MS_PRIVATE which is what we want here, I think.